_posts/2017/2017-12-11-set-title-in-the-view-template-using-content-for.md

```diff
@@ -0,0 +1,69 @@
+---
+title: 'Set your page title in the view template using content_for'
+description: 'Cut out all that instance variable nonsense.'
+layout: article
+category: ruby
+image:
+  base: '2017/set-title-in-the-view-template-using-content-for'
+  alt: 'Books'
+  credit: 'Annie Spratt'
+  source: 'https://unsplash.com/photos/-9vMBjrU-RA'
+---
+
+Rails uses `content_for` as its primary way to store content in one place for use in other views, layouts or helpers. Find out more about `content_for` in the Rails documentation in the [ActionView helpers](http://api.rubyonrails.org/v5.1.3/classes/ActionView/Helpers/CaptureHelper.html#method-i-content_for) section.
+
+## Instead of…
+
+...using an instance variable in the controller...
+
+### `things_controller.rb`
+
+```ruby
+def show
+  @page_title = 'Page Title'
+  # other controller stuff
+end
+```
+
+### And `views/layouts/application.html.erb`
+
+In the `<head>`.
+
+```erb
+<title><%= @page_title || 'Default' %></title>
+```
+
+
+## Use…
+
+...`content_for` and `yield`.
+
+### `things/show.html.erb`
+
+I like to put the title at the top of the individual view file.
+
+```erb
+<% content_for(:html_title) { 'Title' } %>
+```
+
+### And `views/layouts/application.html.erb`
+
+In the `<head>`.
+
+```erb
+<title><%= content_for?(:html_title) ? yield(:page_title) : "Default" %></title>
+```
+
+
+## But why?
+
+Using `content_for` keeps information about how to render the page inside the relevant view template, it avoids cluttering the controller with non-business logic.
+
+You avoid creating yet another 'magic' instance variable that gets automatically passed to the view by Rails.
+
+Here's another way you know I'm right about how you should set your page titles. Rendering JSON or AJAX from the same RESTful controller action doesn't require adding a page title. Therefore, it's clear that this information should live with the relevant HTML view template.
```

Comment:
Colon after 'titles' rather than a full-stop.

Reply:
I get what you're saying...but I'm still not convinced that it is 'clear'...I wonder if there's another way you can say this...

We need something that explains either what it would mean if you did have to set a page title there...or why it's key that you don't have to set a page title there...

Like, I may be reading this sentence — Rendering JSON or AJAX from the same RESTful controller action doesn't require adding a page title. — not understand the significance of that statement so think: 🤷🏾‍♀️...then the next thing you say is: "therefore it's clear..." and I'm thinking: no, it's not..

What the "Rendering JSON or AJAX.." sentence says to me is that it's clear where the info shouldn't go/doesn't need to go...but not clear that it should therefore be in the view.

Do you perhaps want to get across: therefore, Rails is built with the expectation that page title info is in the view hence why you don't need to set one when rendering JSON or AJAX...?

Which bit proves that it still can't live in the controller?

This has been one long ramble...but hopefully you get what I'm saying.

Reply:
How about...

Not convinced? If you were to render a JSON resource from the same controller action, using Rails' `respond_to` [syntax](http://guides.rubyonrails.org/action_controller_overview.html#rendering-xml-and-json-data), those JSON files wouldn't need the content for the HTML `<title>`. Thus the information about the HTML `<title>` _only_ belongs in the HTML template.